source/tutorial/gsfa/GSFA.rst

```diff
@@ -0,0 +1,414 @@
+-  concave points (number of concave portions of the contour)
+-  symmetry
+-  fractal dimension (“coastline approximation” - 1)
+
+*The mean, standard error, and “worst” or largest (mean of the three
+worst/largest values) of these features were computed for each image,
+resulting in 30 features. For instance, field 0 is Mean Radius, field 10
+is Radius SE, field 20 is Worst Radius.*
+
+**Class:**
+
+- WDBC-Malignant
+- WDBC-Benign
+
+
+First let’s import the usual data science modules and library
+`mdp <https://mdp-toolkit.github.io>`__. We use it for GSFA
+implementation and other data processing tools.
+
+
+	>>> import matplotlib.pyplot as plt
+	>>> from sklearn import (datasets, model_selection, metrics)
+	>>> import numpy as np
+	>>> import mdp 
+	>>> from sklearn.svm import SVC
+
+
+Now from the sklearn let’s load the ‘breast_cancer’ dataset. We’ll use
+20% of the data for testing.
+
+
+	>>> data, label = datasets.load_breast_cancer(return_X_y=True
+	>>> data_train, data_test, label_train, label_test = \
+	... model_selection.train_test_split(data, label, test_size=0.2)
+
+We get the baseline classification quality.
+
+	>>> clf = SVC(gamma='auto')
+	>>> clf.fit(data_train, label_train)
+	>>> SVM_test = clf.predict(data_train)
+	>>> print("SVM test score: ", metrics.accuracy_score(label_train, SVM_test))
+
+
+**SVM train score:  1.0**
+
+
+	>>> clf = SVC(gamma='auto')
+	>>> clf.fit(data_train, label_train)
+	>>> SVM_test = clf.predict(data_test)
+	>>> print("SVM test score: ", metrics.accuracy_score(label_test, SVM_test))
+
+
+**SVM test score:  0.5614**
+
+
+Let’s train GSFA model on training data, so that it will compute the
+slowest features possible according to the GSFA optimization problem.
+Since the label information is encoded in the graph connectivity, the
+low-dimensional data are highly predictive of the labels.
+
+We assign the parameter that is responsible for output dimension to 2.
+This will let us to represent the original data in 2D coordinates.
+
+
+	>>> output_dim = 2
+	>>> GSFA_n = mdp.nodes.GSFANode(output_dim=output_dim)
+	>>> GSFA_n.train(data_train, train_mode=("classification", label_train, 1))
+	>>> GSFA_n.stop_training()
+	
+Then we find projections of the train and test data on feature space we have found.
+    
+	>>> GSFA_train = GSFA_n.execute(data_train)
+	>>> GSFA_test = GSFA_n.execute(data_test)
+
+Since we consider dimensionality reduction to the value of 2 we can
+try to depict train and test data on 2-D graph.
+
+
+	>>> import matplotlib
+	>>> colors = ['red', 'blue']
+	>>> f = plt.figure(figsize=(15,6))
+	>>> ax = f.add_subplot(121)
+	>>> ax2 = f.add_subplot(122)
+	>>> ax.scatter(GSFA_train[:, 0], GSFA_train[:, 1], c=label_train,
+	...	cmap=matplotlib.colors.ListedColormap(colors))
+	>>> ax2.scatter(GSFA_test[:, 0], GSFA_test[:, 1], c=label_test,
+	...	cmap=matplotlib.colors.ListedColormap(colors))
+	>>> ax.set_title("Train data in 2-D")
+	>>> ax2.set_title("Test data in 2-D")
+
```

Comment:
@Stewori Should there be a `plt.show()` here?

Reply:
Yes, @nkgevorgyan please add this here. I guess this was forgotten due to porting the code from an ipython notebook.